state/traverse.go

```diff
@@ -7,27 +7,61 @@ import (
 	"github.com/onflow/flow-go/storage"
 )
 
-// Traverse traverses a chain segment beginning with the start block (inclusive)
-// and ending with the end block (inclusive). Blocks are traversed in reverse
+// functor that will be get called on each block header when traversing blocks.
+type onVisitBlock = func(header *flow.Header) (bool, error)
+
+// TraverseBackwards traverses a chain segment beginning with the start block (inclusive)
+// Blocks are traversed in reverse
 // height order, meaning the end block must be an ancestor of the start block.
 // The callback is called for each block in this segment.
-func Traverse(headers storage.Headers, start, end flow.Identifier, callback func(header *flow.Header) error) error {
-
-	nextID := start
+// Return value of callback is used to decide if it should continue or not.
+func TraverseBackwards(headers storage.Headers, startBlockID flow.Identifier, visitor onVisitBlock) error {
+	blockID := startBlockID
 	for {
-		// retrieve the next block in the segment and pass it to the callback
-		next, err := headers.ByBlockID(nextID)
+		block, err := headers.ByBlockID(blockID)
 		if err != nil {
-			return fmt.Errorf("could not get segment block (id=%x): %w", nextID, err)
+			return fmt.Errorf("could not get block header (%x): %w", blockID, err)
+		}
+
+		shouldContinue, err := visitor(block)
+		if !shouldContinue {
+			break
 		}
-		err = callback(next)
 		if err != nil {
-			return fmt.Errorf("error in callback: %w", err)
+			return err
 		}
+		blockID = block.ParentID
+	}
+	return nil
+}
 
-		if nextID == end {
-			return nil
-		}
-		nextID = next.ParentID
+// TraverseForward traverses a chain segment in forward order.
+// Implements a recursive descend to genesis block or till `shouldContinue` returns true and then calls
+// visitor callback for every block way back up to `startBlockID`
+func TraverseForward(headers storage.Headers,
+	startBlockID flow.Identifier,
+	visitor func(header *flow.Header) error,
+	shouldContinue func(header *flow.Header) bool,
+) error {
+	block, err := headers.ByBlockID(startBlockID)
+	if err != nil {
+		return fmt.Errorf("could not get block header (%x): %w", startBlockID, err)
+	}
+
+	if !shouldContinue(block) {
+		return nil
+	}
+
+	// descend further down the chain
+	err = TraverseForward(headers, block.ParentID, visitor, shouldContinue)
```

Comment:
A bit concerned that the recursion , it might cause stackoverflow if the chain is long enough.

I have the impression that it can only support 10K blocks. We can have a test case to verify it

In other words, if we had for instance 10K unsealed blocks to traverse, this function will crash the node. 

A solution would be caching all the nodes in a slice, and using a loop to traverse in a forward order. And slice is unbound, the chain can go as long as the memory can support

Reply:
Implemented with slight modifications.